idaes/generic_models/properties/core/eos/virial.py

```diff
@@ -0,0 +1,797 @@
+                                            f"{b}. Virial equation of"
+                                            "state supports only Vap phase.")
+
+        # get user specified eos options and save on parameter block
+        try:
+            use_pseudocritical_rules = b.config.equation_of_state_options[
+                "use_pseudocritical_rules"]
+            b._use_pseudocritical_rules = use_pseudocritical_rules
+            setattr(b.parent_block(), "virial_eos_options",
+                    deepcopy(VirialConfig))
+            ConfigBlock = getattr(b.parent_block(), "virial_eos_options")
+            ConfigBlock.set_value(b.config.equation_of_state_options)
+        except TypeError:
+            use_pseudocritical_rules = True
+            b._use_pseudocritical_rules = use_pseudocritical_rules
+            setattr(b.parent_block(), "virial_eos_options",
+                    deepcopy(VirialConfig))
+            ConfigBlock = getattr(b.parent_block(), "virial_eos_options")
+            ConfigBlock.set_value(b.config.equation_of_state_options)
+
+        # Get critical volume of components
+        units = b.parent_block().get_metadata().derived_units
+        for comp in b.parent_block().component_list:
+            cobj = b.parent_block().get_component(comp)
+            if not hasattr(cobj, 'volume_crit'):
+                setattr(cobj, "volume_crit",
+                        Var(doc="Pure component critical Volume",
+                            units=units["volume"] / pyunits.mol))
+                set_param_from_config(cobj, param="volume_crit")
+
+        # kappa binary interaction parameters for i-j molecular pair
+        # for  i=j(like molecules), kappa = 0
+        # for i not equal to  j, kappa is a small positive number.
+        # set to zero if data is unavailable
+
+        try:
+            kappa_data = param_block.config.parameter_data["kappa"]
+            param_block.add_component(
+                'kappa',
+                Var(param_block.component_list,
+                    param_block.component_list,
+                    within=NonNegativeReals,
+                    initialize=kappa_data,
+                    doc='Binary interaction parameters',
+                    units=pyunits.dimensionless))
+        except KeyError:
+            param_block.add_component(
+                'kappa',
+                Var(param_block.component_list,
+                    param_block.component_list,
+                    within=NonNegativeReals,
+                    initialize=0.0,
+                    doc='Binary interaction parameters',
+                    units=pyunits.dimensionless))
+
+        # Since standard property changes of formation are readily available for
+        # Enthalpy(H_ref) and Gibbs free energy (G_ref),
+        # This builds G_ref and set its value from config block, and  then
+        # computes S_ref = (H_ref-G_ref)/T_ref
+        for comp in b.parent_block().component_list:
```

Comment:
Again, this is not the right place for this. The `component` modules are responsible for doing this, not the EoS.

Reply:
@ Andrew as a first pass, these are built here but yes, some changes will be required in the component module, on-demand property system and I will suggest we resolved these in a meeting.